torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1239,29 +1239,77 @@ void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
   LOG(FATAL) << logPrefix() << errMsg;
 }
 
+int computeDeltaMS(
+    std::chrono::time_point<std::chrono::steady_clock> start,
+    std::chrono::time_point<std::chrono::steady_clock> end) {
+  return std::chrono::duration_cast<std::chrono::milliseconds>(end - start)
+      .count();
+}
+
 void ProcessGroupNCCL::heartbeatMonitor() {
   uint64_t heartBeatCounter = 0ULL;
+  bool checkTimeoutSignal = (dumpOnTimeout_ && uid_ == 0);
+  int heartBeatTimeout = checkTimeoutSignal ? coordDumpCheckIntervalMilSec_
+                                            : heartbeatTimeoutInSec_ * 1000;
+  auto lastTimePollStore = std::chrono::steady_clock::now();
+  auto lastTimeHeartBeatCheck = std::chrono::steady_clock::now();
+  std::future<bool> asyncDebugDump;
   while (true) {
     // This won't have any lock since this lock is only used here.
     // Please be aware that mutex `monitorMutex_` should not be used
     // somewhere else to avoid the deadlock.
     std::unique_lock<std::mutex> lock(monitorMutex_);
     if (monitorWakeUpCV_.wait_for(
-            lock, std::chrono::seconds(heartbeatTimeoutInSec_), [&] {
+            lock, std::chrono::milliseconds(heartBeatTimeout), [&] {
               return terminateHeartbeatMonitorThread_.load();
             })) {
       // For the normal complete or user interception, monitorWakeUpCV_
       // will get notified, we early return and exit heartbeatMonitor.
       return;
     }
+    auto currentTime = std::chrono::steady_clock::now();
+
+    // We put extra functionality in the thread for the default PG (aka, uid_=0)
+    // because the signal is same across different PGs. We only need to run
+    // once per process to avoid duplicate things performed in too many separate
+    // threads. For example, we check a global flag on the TCPStore periodically
+    // to see if any PG on any rank observed a timeout and signaled peers to
+    // dump debugging info, and we avoid hammering the TCPStore from all PGs on
+    // the same rank.
+    if (checkTimeoutSignal) {
+      // We poll store to see if some ranks have flagged a timeout when
+      // we haven't polled for `heartbeat_timeout` seconds and there haven't
+      // any work added or removed for `watchdog_timeout` seconds.
+      if (computeDeltaMS(lastWorkListUpdateTime_, currentTime) >=
+              kWatchdogThreadSleepMillis &&
+          computeDeltaMS(lastTimePollStore, currentTime) >=
+              coordDumpCheckIntervalMilSec_) {
+        lastTimePollStore = currentTime;
+        if (globalStore_->check({std::string(TIMEOUT_DUMP)})) {
+          auto wakeUpTime = getWakeupTime(waitTimeoutDumpInMilSec_);
+          asyncDebugDump = launchAsyncDebugDump();
```

Comment:
do we need to have the code for launching/dumping inside this if?

why not just 'break' here and rely on the normal exit path below to do the dump?

We can add one log here saying "received dump signal" so we can tell that that's why we dumped rather than actual heartbeat failure.

We can also add a log in the heartbeat comparison saying "heartbeat missed"

or we can create an empty string at the top, set it to either "dump signaled by other rank", or "heartbeat not detected"

then in the codepath below we can just append this "reason" log to the main log we emit

Reply:
Hmmm, but that would lead to program aborted instead of exception throwing. We have no exception throwing anyway. Maybe this should be fine?

Reply:
if you want to throw in one case and abort in other case, can you say why?  (or should we just be consistent?)

Reply:
Well.. you are right, this exception is a legacy from watchdog where we only throw exception while in monitoring thread, we just do abort. But I don't know if switching to abort will make it testable or not. Let me try.

Reply:
Oh Abort actually works for unit test, nice! OK I think I can make this change.